pkg/controller/sparkapplication/controller.go

```diff
@@ -289,386 +250,461 @@ func (c *Controller) processNextItem() bool {
+//|      +----+----+    |    +-----v----+          +----------+           +-----------+          +----v-----+          |
+//|      |         |    |    |          |          |          |           |           |          |          |          |
+//|      |         |    |    |          |          |          |           |           |          |			|          |
+//|      |   New   +---------> Submitted+----------> Running  +----------->  Failing  +---------->  Failed  |          |
+//|      |         |    |    |          |          |          |           |           |          |          |          |
+//|      |         |    |    |          |          |          |           |           |          |          |          |
+//|      |         |    |    |          |          |          |           |           |          |          |          |
+//|      +---------+    |    +----^-----+          +-----+----+           +-----+-----+          +----------+          |
+//|                     |         |                      |                      |                                      |
+//|                     |         |                      |                      |                                      |
+//|                     |         |             +-------------------------------+                                      |
+//|                     |   +-----+-----+       |        |                +-----------+          +----------+          |
+//|                     |   |  Pending  |       |        |                |           |          |          |          |
+//|                     +---+   Retry   <-------+        +---------------->Succeeding +---------->Completed |          |
+//|                         |           <-------+                         |           |          |          |          |
+//|                         |           |       |                         |           |          |          |          |
+//|                         |           |       |                         |           |          |          |          |
+//|                         +-----------+       |                         +-----+-----+          +----------+          |
+//|                                             |                               |                                      |
+//|                                             |                               |                                      |
+//|                                             +-------------------------------+                                      |
+//|                                                                                                                    |
+//+--------------------------------------------------------------------------------------------------------------------+
+func (c *Controller) syncSparkApplication(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
```

Comment:
This method now looks really long and hard to review. Can we break it down into smaller functions, e.g., putting the logic on a per-state basis into individual helper functions?